Clean up location table

Pepe20129 opened the pull request:
Cleans up some redundant stuff in the location table (categories, location type & check type were doing basically the same thing for example).

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh.otr.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2000978217.zip)
  - [soh-linux.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2001002168.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2001011016.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/2001022921.zip)
<!--- section:artifacts:end -->

Pepe20129 committed:
Remove categories

Pepe20129 committed:
Remove SpoilerCollectionCheckGroup

Pepe20129 committed:
Remove unused Location constructors

Pepe20129 committed:
Remove LocationType

Pepe20129 committed:
Remove unused SpoilerCollectionCheckTypes

Pepe20129 committed:
Remove hints' isVanillaCompletion, it's always false

Pepe20129 committed:
Remove SpoilerCollectionCheck::None

Same effect as default constructor

Pepe20129 committed:
Fix mistake

Pepe20129 committed:
Merge branch 'develop-rando' into clean_up_location_table

boomshroom commented:
The `flags` field also seems to be redundant and anything that would use it accesses it through `GetCollectionCheck().flag`.

Pepe20129 committed:
Update location.h

Pepe20129 committed:
Merge branch 'develop-rando' into clean_up_location_table

Pepe20129 committed:
Deduce RC Area by scene when possible

Pepe20129 committed:
Remove useless constructors

Pepe20129 committed:
Remove flag in most situations

Pepe20129 committed:
Format improvements

Pepe20129 committed:
Replace overworldLocations with a function

Pepe20129 committed:
Replace shopLocationLists with a function

Pepe20129 committed:
Replace gossipStoneLocations with a function

Also add a new RCType for static hints

Pepe20129 committed:
Replace scrubLocations with a function

Pepe20129 committed:
Replace staticHintLocations with a function

Pepe20129 committed:
Replace overworldFishLocations with a function

Pepe20129 committed:
Replace pondFishLocations with a function

Pepe20129 committed:
Merge branch 'develop-rando' into clean_up_location_table

Pepper0ni commented on soh/soh/Enhancements/randomizer/context.cpp:
```diff
@@ -140,7 +140,7 @@ void Context::GenerateLocationPool() {
         AddLocation(RC_TRIFORCE_COMPLETED);
     }
-    AddLocations(StaticData::overworldLocations);
+    AddLocations(StaticData::GetOverworldLocations());
```
This function erroneously includes RC_TRIFORCE_COMPLETED in it's search, causing it to be added when triforce hunt is off, causing seeds to fail to generate without it if entrance rando is on due to the stricter AllLocationsAccessible check.

Pepe20129 replied:
Done.

Pepe20129 committed:
Remove RC_TRIFORCE_COMPLETED from GetOverworldLocations

Pepper0ni commented on soh/soh/Enhancements/randomizer/context.cpp:
```diff
@@ -140,7 +140,7 @@ void Context::GenerateLocationPool() {
-    AddLocations(StaticData::overworldLocations);
+    AddLocations(StaticData::GetOverworldLocations());
 
     if (mSettings->GetOption(RSK_FISHSANITY).IsNot(RO_FISHSANITY_OFF)) {
```
Fishsanity checks are added when it is off, likely because of GetOverworldLocations catching those too. Trying to play with fishanity on both is also crashing but I don't know if that's the same issue or not

Pepe20129 replied:
In that case we might want to change it so that we don't add all overworld locations and instead we add a filter to only add some overworld checks.

Pepper0ni replied:
One of my "Sould probably get done eventually" things was changing it so only enabled checks are in the location pool at all, instead of adding everything and then defaulting the unused things. Originally it was just keeping the spoiler log clean but it would be better than this wierd peicemeal approach. This could also simplify building the item pool if we simply add the vanilla checks of those locations as we go. Perhaps a good starting point would be adding a condition to add the check to the pool to each location.

Malkierian replied:
Has this been resolved, or is it going to be a fix later thing?

Pepe20129 replied:
Probably a fix later thing as this PR is already too big.

garrettjoecox reviewed (commented):
I support the removal of the duplicated data from the location table, though my thought was to merge the SpoilerCollectionCheck stuff into the table itself, is there a reason for us to store that on a child class instance, rather than just storing the flag type and flag on the table?

Pepe20129 commented:
Long term, no, but currently I think it's better to leave it like this as this PR is already too big and there's not much consistency on what uses it and what doesn't so I think leaving it like this can make it easier to see in the future as an area of improvement.

Malkierian commented:
Resolve conflicts with this one and I'll look at it for the next merge.

Pepe20129 committed:
Merge branch 'develop-rando' into clean_up_location_table

Pepe20129 commented:
Fixed the merge conflicts.

Malkierian commented on soh/soh/Enhancements/randomizer/3drando/fill.cpp:
```diff
@@ -979,10 +979,7 @@ static void RandomizeOwnDungeon(const Rando::DungeonInfo* dungeon) {
-      return !(Rando::StaticData::GetLocation(loc)->IsCategory(Category::cSong));
-    }
-    if (ctx->GetOption(RSK_SHUFFLE_SONGS).Is(RO_SONG_SHUFFLE_SONG_LOCATIONS)) {
-      return !(Rando::StaticData::GetLocation(loc)->IsCategory(Category::cSongDungeonReward));
```
This was combining things like Lens of Truth chest before in the grouping. The changes don't give those locations `RCTYPE_SONG_LOCATION`, so they're no longer included. How does this affect generation? Was it a mistake to include them here to begin with?

Pepe20129 replied:
That's for `RO_SONG_SHUFFLE_DUNGEON_REWARDS`, it was a mistake to not include them anymore.
Although idk why it was checking `RO_SONG_SHUFFLE_SONG_LOCATIONS`, that was probably wrong before.

Malkierian commented on soh/soh/Enhancements/randomizer/3drando/fill.cpp:
```diff
@@ -1246,11 +1242,24 @@ int Fill() {
+                   loc == RC_SPIRIT_TEMPLE_TWINROVA_HEART ||
+                   loc == RC_SHADOW_TEMPLE_BONGO_BONGO_HEART ||
+                   loc == RC_SHEIK_IN_ICE_CAVERN ||
+                   loc == RC_SONG_FROM_IMPA;
```
It looks like these all share the same RCTYPE, `RCTYPE_BOSS_HEART_OR_OTHER_REWARD` except the two song locations. Could probably simplify the check that way. Also, why only the two song locations?

Pepe20129 replied:
It's for `RO_SONG_SHUFFLE_DUNGEON_REWARDS`, which puts songs at "Dungeon Reward Locations", there's 2 more songs than those locations so 2 regular song locations are also included.

Malkierian replied:
So the Song Shuffle "Dungeon Reward" locations aren't the same as the Dungeon Rewards "End of Dungeons"? That probably should be clarified in the tooltips... Still, would probably do the check for RCTYPE as well as RC for the two song locations.

Malkierian commented on soh/soh/Enhancements/randomizer/3drando/spoiler_log.cpp:
```diff
@@ -99,12 +99,12 @@ void WriteIngameSpoilerLog() {
-        return groupA < groupB;
+        RandomizerCheckArea areaA = Rando::StaticData::GetLocation(a)->GetArea();
+        RandomizerCheckArea areaB = Rando::StaticData::GetLocation(b)->GetArea();
+        return areaA < areaB;
```
How comparable were RCA and CCG in this regard?

Pepe20129 replied:
Pretty similar except for some small differences like dmt & dmc being together (iirc market + tot + castle grounds were also combined into 1).

Malkierian replied:
So, SCCG actually had fewer groups than RCA does? I can live with that, as long as it doesn't affect other parts of logic. @Pepper0ni would that change anything about hint generation, fundamentally?

Pepper0ni replied:
I literally deleted this whole function in my shop unification as it's an unused 3DS feature.

Malkierian replied:
Right, but this isn't the only place SCCG was replaced by RCA. This was just where I first saw it.

Malkierian reviewed (commented):
Just a few things, mostly questions.

Malkierian committed:
Update fill.cpp

Re-add filter for dungeonLocations for song shuffle dungeon rewards (and fix erroneous check for song locations there).
Modify songLocations filter for dungeon rewards to check for RCTYPE_BOSS_HEART_OR_OTHER_REWARD and then the two song locations added for it.

Malkierian committed:
Update fill.cpp

Fix bracketing.

Pepe20129 committed:
Apply Pepper0ni's patch

Pepper0ni reviewed (approved):
This is going to need some forward fixes, but getting things in and fixing them after is the way forward
